### ISSUE TITLE
type checking on boolean expressions

### ISSUE THREAD

**Issue by TadLeonard** [python/mypy#1871]
I came across this while reading [Daniel Moisset's impression](http://www.machinalis.com/blog/a-day-with-mypy-part-2/) of mypy. He mentions satisfying mypy by wrapping a boolean return expression in `bool()`, which I think is unnecessary. Below is a simple example that mypy fails on with `Incompatible return value type (got "Union[str, bool]", expected "bool")`.  Note that the return expression is of the form `<expression> and <comparison> or <comparison>`. 

```python
def bool_expr_test(line: str) -> bool:
    return line and line[0] == '"' or line[0] == "'"
```

Tested on 64c0acd4e10e7ba443b1fffba6e945e6943cd6ea and 0.4.3 (from `pip`).

**Comment by gvanrossum** [comment]
See also #1825

--Guido (mobile)

**Comment by gvanrossum** [comment]
However, 'line and True' does not always return bool. If line is '' it returns that.

**Comment by rwbarton** [comment]
Also see #1477 for the first example. The remaining examples are all correct as @gvanrossum noted (`line` could be `''` or `0`).

**Comment by TadLeonard** [comment]
Ah, thanks. Removed simpler `line and ` examples from the ticket. I thought I knew Python, but I guess I usually use `<expr> or <expr>` as assignment and return expressions. Never thought to use `<expr> and <expr>`.

**Comment by gvanrossum** [comment]
There's an interesting symmetry though -- *in general* both `<expr> or
<expr>` and `<expr> or <expr>` can return the type of either the first or
the second expression. It's mostly when you combine them that interesting
and wonderful things happen. You can probably construct an example using
`(<expr> or <expr>) and <expr>` too. :-)

**Comment by ddfisher** [comment]
This will be fixed by #1751 (which was partially implemented by #1828).

**Comment by ddfisher** [comment]
Actually, I think #1751 isn't sufficient to fix this.  It looks like we'll need #1698.

**Comment by gvanrossum** [comment]
Here's another example I ran into: https://github.com/python/mypy/blob/c528638891bffcdb53f526e06053ec768dfe10fb/mypy/main.py#L342-L343
```
mypy/main.py:343: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")
```
The error is misguided: while `strip_py(mod)` indeed returns `Optional[str]`, the `or mod` turns the `None` into a `str` and hence the type of `strip_py(mod) or mod` is in fact `str`.

**Comment by ddfisher** [comment]
It's odd that that doesn't work, because we actually have an [explicit test for that behavior](https://github.com/python/mypy/blob/c528638891bffcdb53f526e06053ec768dfe10fb/test-data/unit/check-optional.test#L179).

**Comment by gvanrossum** [comment]
The test has a contextthat may affect it. This definitely gives an error
with strict optional:
```python
def f(s: str) -> Optional[str]: pass
x = ''
x = f(x) or x
```

**Comment by rwbarton** [comment]
Oh, the difference is because the logic for this case goes through the type binder, which only stores information about the types of expressions like `x[i].a`, not expressions like `f(x)`. (It's originally intended for handling things like `if isinstance(x[i].a, C): ... x[i].a ...`, where there are two occurrences of the expression `x[i].a`. We assume that `x[i].a` will not change type in ways that are invisible to the binder, but we don't assume that about function calls like `f(x)`.)

**Comment by gvanrossum** [comment]
So what should we do?

**Comment by dmoisset** [comment]
Both these examples (`bool_expr_test` on the ticket description, and the later one posted by @gvanrossum ) seem to work ok and fixed after applying #1989

**Comment by ddfisher** [comment]
Fixed by #1989.
